core/domain/exp_domain.py

```diff
@@ -3060,90 +3169,1468 @@ def _convert_states_v51_dict_to_v52_dict(
+                    # strings is a substring of the other contains
+                    # rule strings.
+                    for contain_rule_ele in seen_strings_contains:
+                        for contain_rule_string in contain_rule_ele:
+                            for rule_value in rule_values:
+                                if contain_rule_string in rule_value:
+                                    invalid_rules.append(rule_spec)
+                    seen_strings_contains.append(rule_values)
+                elif rule_spec['rule_type'] == 'StartsWith':
+                    # `StartsWith` rule should always come after another
+                    # `StartsWith` rule where the first starts-with string
+                    # is the prefix of the other starts-with string.
+                    for start_with_rule_ele in seen_strings_startswith:
+                        for start_with_rule_string in start_with_rule_ele:
+                            for rule_value in rule_values:
+                                if rule_value.startswith(
+                                    start_with_rule_string):
+                                    invalid_rules.append(rule_spec)
+                    # `Contains` should always come after `StartsWith` rule
+                    # where the contains rule strings is a substring
+                    # of the `StartsWith` rule string.
+                    for contain_rule_ele in seen_strings_contains:
+                        for contain_rule_string in contain_rule_ele:
+                            for rule_value in rule_values:
+                                if contain_rule_string in rule_value:
+                                    invalid_rules.append(rule_spec)
+                    seen_strings_startswith.append(rule_values)
+                elif rule_spec['rule_type'] == 'Equals':
+                    # `Contains` should always come after `Equals` rule
+                    # where the contains rule strings is a substring
+                    # of the `Equals` rule string.
+                    for contain_rule_ele in seen_strings_contains:
+                        for contain_rule_string in contain_rule_ele:
+                            for rule_value in rule_values:
+                                if contain_rule_string in rule_value:
+                                    invalid_rules.append(rule_spec)
+                    # `Startswith` should always come after the `Equals`
+                    # rule where a `starts-with` string is a prefix of the
+                    # `Equals` rule's string.
+                    for start_with_rule_ele in seen_strings_startswith:
+                        for start_with_rule_string in start_with_rule_ele:
+                            for rule_value in rule_values:
+                                if rule_value.startswith(
+                                    start_with_rule_string):
+                                    invalid_rules.append(rule_spec)
+
```

Comment:
Shouldn't `seen_strings_startswith.append(rule_values)` be added here?

Reply:
Nope, this part is present inside the `Equals` rule and we add the `StartsWith` rule in the above part.